packages/gatsby/src/redux/reducers/redirects.ts

```diff
@@ -0,0 +1,49 @@
+import _ from "lodash"
+
+import { IGatsbyState, IRedirect, ICreateRedirectAction } from "../types"
+
+const redirects = new Map<string, IRedirect[]>()
+
+function exists(newRedirect: IRedirect): boolean {
+  const fromPathRedirects = redirects.get(newRedirect.fromPath)
+
+  if (!fromPathRedirects) return false
+
+  return fromPathRedirects.some(redirect => _.isEqual(redirect, newRedirect))
+}
+
+function add(redirect: IRedirect): void {
+  let samePathRedirects = redirects.get(redirect.fromPath)
+
+  if (!samePathRedirects) {
+    samePathRedirects = []
+    redirects.set(redirect.fromPath, samePathRedirects)
+  }
+
+  samePathRedirects.push(redirect)
+}
+
+const redirectsReducer = (
+  state: IGatsbyState["redirects"] = [],
+  action: ICreateRedirectAction
+): IGatsbyState["redirects"] => {
+  switch (action.type) {
+    case `CREATE_REDIRECT`: {
+      const redirect = action.payload
+
+      // Add redirect only if it wasn't yet added to prevent duplicates
+      if (!exists(redirect)) {
+        add(redirect)
+
+        state.push(redirect)
+      }
+
+      return state
+    }
+
+    default:
+      return state
+  }
+}
+
+export default redirectsReducer
```

Comment:
We want to have named exports as per https://github.com/gatsbyjs/gatsby/issues/21995